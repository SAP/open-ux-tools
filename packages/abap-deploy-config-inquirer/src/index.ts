import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
