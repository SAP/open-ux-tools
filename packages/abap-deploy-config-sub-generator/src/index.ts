export { getAbapQuestions } from './app/questions';
export { indexHtmlExists, showOverwriteQuestion } from './utils';
export { AbapDeployConfigOptions, DeployProjectType } from './app/types';
