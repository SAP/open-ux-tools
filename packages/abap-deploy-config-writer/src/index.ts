import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import cloneDeep from 'lodash/cloneDeep';
import { addPackageDevDependency, FileName, readUi5Yaml, updatePackageScript } from '@sap-ux/project-access';
import { getDeployConfig, updateBaseConfig } from './config';
import { addUi5Dependency, writeUi5RepositoryFiles } from './file';
import {
    BUILD_SCRIPT,
    DEPLOY_SCRIPT,
    RIMRAF,
    RIMRAF_VERSION,
    UI5_TASK_FLATTEN_LIB,
    UI5_TASK_FLATTEN_LIB_VERSION,
    UNDEPLOY_SCRIPT
} from './constants';
import type { DeployConfigOptions } from './types';
import type { Editor } from 'mem-fs-editor';
import type { AbapDeployConfig } from '@sap-ux/ui5-config';

/**
 * Writes the template to the memfs editor instance.
 *
 * @param basePath - the base path
 * @param abapDeployConfig - the abap deploy config
 * @param options - the deploy config options
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
async function generate(
    basePath: string,
    abapDeployConfig: AbapDeployConfig,
    options?: DeployConfigOptions,
    fs?: Editor
): Promise<Editor> {
    const abapConfig = cloneDeep(abapDeployConfig);

    if (!fs) {
        fs = create(createStorage());
    }

    // base config
    const baseConfigFile = options?.baseFile ?? FileName.Ui5Yaml;
    const baseConfig = await readUi5Yaml(basePath, baseConfigFile, fs);
    const baseConfigPath = join(basePath, baseConfigFile);
    updateBaseConfig(baseConfigPath, baseConfig, fs);

    // deploy config
    const deployConfigFile = options?.deployFile ?? FileName.UI5DeployYaml;
    const deployFilePath = join(basePath, deployConfigFile);
    const deployConfig = await getDeployConfig(abapConfig, baseConfig);
    fs.write(deployFilePath, deployConfig.toString());

    // package.json
    // deploy script
    const deployScript = `${BUILD_SCRIPT} && ${DEPLOY_SCRIPT} --config ${deployConfigFile} && ${RIMRAF} archive.zip`;
    await updatePackageScript(basePath, 'deploy', deployScript, fs);
    // undeploy script
    const undeployScript = `${BUILD_SCRIPT} && ${UNDEPLOY_SCRIPT} --config ${deployConfigFile}`;
    await updatePackageScript(basePath, 'undeploy', undeployScript, fs);
    // test mode script
    const deployTestModeScript = `${BUILD_SCRIPT} && ${DEPLOY_SCRIPT} --config ${deployConfigFile} --testMode true`;
    await updatePackageScript(basePath, 'deploy-test', deployTestModeScript, fs);
    // dependencies
    await addPackageDevDependency(basePath, RIMRAF, RIMRAF_VERSION, fs);
    await addPackageDevDependency(basePath, UI5_TASK_FLATTEN_LIB, UI5_TASK_FLATTEN_LIB_VERSION, fs);
    addUi5Dependency(fs, basePath, UI5_TASK_FLATTEN_LIB);

    // ui5 repo files
    await writeUi5RepositoryFiles(fs, basePath);

    return fs;
}

export { generate };
