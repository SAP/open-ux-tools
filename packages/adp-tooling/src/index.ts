export { AdpPreviewConfig, AdpWriterConfig } from './types';
export * from './preview/adp-preview';
export { generate } from './writer';
export * from './base/prompt';
