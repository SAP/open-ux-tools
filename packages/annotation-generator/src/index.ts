export { AnnotationServiceParameters, GenerateAnnotationsOptions, generateAnnotations } from './generation';
