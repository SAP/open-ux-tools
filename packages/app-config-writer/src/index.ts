export { getSmartLinksTargetFromPrompt } from './prompt';
export { generateSmartLinksConfig } from './smartlinks-config';
