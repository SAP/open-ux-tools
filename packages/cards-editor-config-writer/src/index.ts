import { join } from 'path';
import type { Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import type { Package } from '@sap-ux/project-access';
import { UI5Config } from '@sap-ux/ui5-config';

/**
 * Check if package.json has devDependency to @ui5/cli version  > 2.
 *
 * @param devDependencies - parsed devDependencies from package.json
 * @returns - true: @ui/cli higher version 2; false: @ui/cli
 */
function isUi5CliHigherTwo(devDependencies: Partial<Record<string, string>>): boolean {
    let isHigherTwo = false;
    try {
        const versionString = devDependencies['@ui5/cli'];
        if (typeof versionString === 'string') {
            const regex = /\d+/;
            const result = regex.exec(versionString.split('.')[0]);
            const majorVersionValue = result ? result[0] : '0';
            const majorVersion = parseInt(majorVersionValue, 10);
            isHigherTwo = majorVersion > 2;
        }
    } catch {
        // if something went wrong we don't have @ui/cli > version 2
    }
    return isHigherTwo;
}

/**
 * Updates the package.json file to include the cards editor middleware and the start-cards-generator script.
 *
 * @param basePath - The path to the project root
 * @param fs - Mem-fs editor instance
 */
function updatePackageJson(basePath: string, fs: Editor) {
    const packageJsonPath = join(basePath, 'package.json');
    if (!fs.exists(packageJsonPath)) {
        throw new Error('package.json not found');
    }

    const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;

    packageJson.devDependencies ??= {};
    packageJson.devDependencies['@sap-ux/cards-editor-middleware'] ??= '0';

    packageJson.scripts ??= {};
    packageJson.scripts['start-cards-generator'] = `fiori run --open 'test/flpGeneratorSandbox.html#Cards-generator'`;

    if (!isUi5CliHigherTwo(packageJson.devDependencies)) {
        packageJson.ui5 ??= {};
        packageJson.ui5.dependencies ??= [];
        if (!packageJson.ui5.dependencies.includes('@sap-ux/cards-editor-middleware')) {
            packageJson.ui5.dependencies.push('@sap-ux/cards-editor-middleware');
        }
    }
    fs.writeJSON(packageJsonPath, packageJson);
}

/**
 * Updates the ui5.yaml file to include the cards editor middleware.
 *
 * @param basePath - The path to the project root
 * @param fs - Mem-fs editor instance
 * @param middlewares - The middlewares to add to the ui5.yaml file
 */
async function updateYaml(basePath: string, fs: Editor, middlewares: string[]) {
    const ui5ConfigPath = join(basePath, 'ui5.yaml');
    if (!fs.exists(ui5ConfigPath)) {
        throw new Error('ui5.yaml not found');
    }
    const config = await UI5Config.newInstance(fs.read(ui5ConfigPath));
    config.addCustomMiddleware(
        middlewares
            .filter((name) => !config.findCustomMiddleware(name))
            .map((name) => ({
                name,
                afterMiddleware: 'compression',
                configuration: undefined
            }))
    );
    fs.write(ui5ConfigPath, config.toString());
}

/**
 * Enables the cards editor in the given project.
 *
 * @param basePath - path to the project root
 * @param fs - optional mem-fs editor instance
 * @returns updated mem-fs editor instance
 */
export async function enableCardsEditor(basePath: string, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }

    updatePackageJson(basePath, fs);
    await updateYaml(basePath, fs, ['sap-cards-generator']);

    return fs;
}
