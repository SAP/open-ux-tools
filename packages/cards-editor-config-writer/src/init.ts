#!/usr/bin/env node

import { enableCardsEditor } from '.';
import { ToolsLogger } from '@sap-ux/logger';

const logger = new ToolsLogger();
enableCardsEditor(process.cwd())
    .then((fs) => {
        fs.commit(() => logger.info('Done'));
    })
    .catch((err) => {
        logger.error(err);
    });
