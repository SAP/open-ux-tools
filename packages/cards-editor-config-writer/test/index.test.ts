import { enableCardsEditor } from '../src';
import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';

function createTestFs(basePath: string) {
    const fs = create(createStorage());
    fs.writeJSON(join(basePath, 'webapp/manifest.json'), {
        'sap.app': {
            id: 'test.id',
            title: 'Test App'
        }
    });
    fs.writeJSON(join(basePath, 'package.json'), {});
    fs.write(join(basePath, 'ui5.yaml'), '');
    return fs;
}

describe('enableCardEditor', () => {
    test('Valid LROP', async () => {
        const basePath = join(__dirname, 'fixtures/lrop-v4');
        const fs = createTestFs(basePath);
        await enableCardsEditor(basePath, fs);

        expect(fs.read(join(basePath, 'package.json'))).toMatchSnapshot();
        expect(fs.read(join(basePath, 'ui5.yaml'))).toMatchSnapshot();
    });

    test('V4 LROP with CLI 3.0', async () => {
        const basePath = join(__dirname, 'fixtures/lrop-v4');
        const fs = create(createStorage());
        await enableCardsEditor(basePath, fs);

        if (process.env.UX_DEBUG) {
            fs.commit(() => {});
        }

        expect(fs.read(join(basePath, 'package.json'))).toMatchSnapshot();
        expect(fs.read(join(basePath, 'ui5.yaml'))).toMatchSnapshot();
    });
});
