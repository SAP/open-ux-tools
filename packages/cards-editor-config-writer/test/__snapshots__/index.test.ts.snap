// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enableCardEditor V4 LROP with CLI 3.0 1`] = `
"{
  \\"name\\": \\"apps.v4.example\\",
  \\"version\\": \\"0.0.1\\",
  \\"private\\": true,
  \\"sapux\\": true,
  \\"description\\": \\"My LROP V4 Application\\",
  \\"scripts\\": {
    \\"build\\": \\"ui5 build\\",
    \\"start\\": \\"ui5 serve --open \\\\\\"/test/flpSandbox.html#app-preview\\\\\\"\\",
    \\"start-cards-generator\\": \\"fiori run --open 'test/flpGeneratorSandbox.html#Cards-generator'\\"
  },
  \\"devDependencies\\": {
    \\"@ui5/cli\\": \\"3.9.0\\",
    \\"@sap/ux-ui5-tooling\\": \\"1\\",
    \\"@sap-ux/ui5-middleware-fe-mockserver\\": \\"2.2.19\\",
    \\"@sap-ux/preview-middleware\\": \\"0.11.26\\",
    \\"@sap-ux/cards-editor-middleware\\": \\"0\\"
  }
}
"
`;

exports[`enableCardEditor V4 LROP with CLI 3.0 2`] = `
"specVersion: \\"2.4\\"
metadata:
  name: apps.v4.example
type: application
server:
  customMiddleware:
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: test/flpSandbox.html
    - name: sap-fe-mockserver
      afterMiddleware: compression
      configuration:
        services:
          - urlPath: '/sap/opu/odata4/mock'
            metadataPath: './webapp/localService/metadata.xml'
            generateMockData: true
    - name: fiori-tools-proxy
      afterMiddleware: sap-fe-mockserver
      configuration:
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
    - name: sap-cards-generator
      afterMiddleware: compression
"
`;

exports[`enableCardEditor Valid LROP 1`] = `
"{
  \\"devDependencies\\": {
    \\"@sap-ux/cards-editor-middleware\\": \\"0\\"
  },
  \\"scripts\\": {
    \\"start-cards-generator\\": \\"fiori run --open 'test/flpGeneratorSandbox.html#Cards-generator'\\"
  },
  \\"ui5\\": {
    \\"dependencies\\": [
      \\"@sap-ux/cards-editor-middleware\\"
    ]
  }
}
"
`;

exports[`enableCardEditor Valid LROP 2`] = `
"server:
  customMiddleware:
    - name: sap-cards-generator
      afterMiddleware: compression
"
`;
